ext/uri/php_uri_common.c

```diff
@@ -0,0 +1,145 @@
+}
+
+void throw_invalid_uri_exception(const uri_handler_t *uri_handler, zval *errors)
+{
+	zval exception_zv;
+
+	uri_handler->create_invalid_uri_exception(&exception_zv, errors);
+
+	zend_throw_exception_object(&exception_zv);
+}
+
+void uri_read_component(INTERNAL_FUNCTION_PARAMETERS, zend_string *property_name, uri_component_read_mode_t component_read_mode)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	uri_internal_t *internal_uri = Z_URI_INTERNAL_P(ZEND_THIS);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	const uri_property_handler_t *property_handler = uri_property_handler_from_internal_uri(internal_uri, property_name);
+	ZEND_ASSERT(property_handler != NULL);
+
+	if (UNEXPECTED(property_handler->read_func(internal_uri, component_read_mode, return_value) == FAILURE)) {
+		zend_throw_error(NULL, "%s::$%s property cannot be retrieved", ZSTR_VAL(Z_OBJ_P(ZEND_THIS)->ce->name), ZSTR_VAL(property_name));
+		RETURN_THROWS();
+	}
+}
+
+static void uri_write_component_ex(INTERNAL_FUNCTION_PARAMETERS, zend_string *property_name, zval *property_zv)
+{
+	uri_internal_t *internal_uri = Z_URI_INTERNAL_P(ZEND_THIS);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	const uri_property_handler_t *property_handler = uri_property_handler_from_internal_uri(internal_uri, property_name);
+	ZEND_ASSERT(property_handler != NULL);
+
+	zend_object *new_object = uri_clone_obj_handler(Z_OBJ_P(ZEND_THIS));
+	if (UNEXPECTED(EG(exception) != NULL)) {
+		zend_object_release(new_object);
+		zval_ptr_dtor(property_zv);
+		RETURN_THROWS();
+	}
+
+	uri_internal_t *new_internal_uri = uri_internal_from_obj(new_object);
+	URI_ASSERT_INITIALIZATION(new_internal_uri);
+	if (property_handler->write_func == NULL) {
+		zend_readonly_property_modification_error_ex(ZSTR_VAL(Z_OBJ_P(ZEND_THIS)->ce->name), ZSTR_VAL(property_name));
+		zend_object_release(new_object);
+		zval_ptr_dtor(property_zv);
+		RETURN_THROWS();
+	}
+
+	zval errors;
+	ZVAL_UNDEF(&errors);
+	if (property_handler->write_func(new_internal_uri, property_zv, &errors) == FAILURE) {
+		throw_invalid_uri_exception(new_internal_uri->handler, &errors);
+		zval_ptr_dtor(&errors);
+		zend_object_release(new_object);
+		zval_ptr_dtor(property_zv);
+		RETURN_THROWS();
+	}
+
+	ZEND_ASSERT(Z_ISUNDEF(errors));
+	RETVAL_OBJ(new_object);
+	zval_ptr_dtor(property_zv);
+}
+
+void uri_write_component_str(INTERNAL_FUNCTION_PARAMETERS, zend_string *property_name)
+{
+	zend_string *value;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_PATH_STR(value)
+	ZEND_PARSE_PARAMETERS_END();
+
+	zval zv;
+	ZVAL_STR_COPY(&zv, value);
+
+	uri_write_component_ex(INTERNAL_FUNCTION_PARAM_PASSTHRU, property_name, &zv);
+}
+
+void uri_write_component_str_or_null(INTERNAL_FUNCTION_PARAMETERS, zend_string *property_name)
+{
+	zend_string *value;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_PATH_STR_OR_NULL(value)
+	ZEND_PARSE_PARAMETERS_END();
+
+	zval zv;
+	if (value == NULL) {
+		ZVAL_NULL(&zv);
+	} else {
+		ZVAL_STR_COPY(&zv, value);
```

Comment:
This copy here looks unnecessary. The value is destructed in all paths in `uri_write_component_ex`:

```suggestion
		ZVAL_STR(&zv, value);
```

(same in the other functions)

Reply:
Good catch! I think I did this because I wanted to pass a string which is safe to own by the URI implementations, but then the problem can be addressed at the `write_handler` level for sure by copying the string there.